FEP-0005: Dependency and Platform Policy

hyarion opened the pull request:
While we have an informal practice of maintaining compatibility with older Ubuntu LTS releases, this has not been formally documented. This proposal aims to address that by establishing it as a formal policy.

It also introduces a process for updating dependencies and supported Ubuntu LTS versions when necessary. The goal is to minimize disruptions for both packagers and developers. We've encountered several instances where changes to dependencies were made without prior notice, leading to broken conda and PPA builds for several weeks.

Suggestions for improvement are welcome.

hyarion committed:
Create FEP-XXXX-process.md

Add dependency and platform policy

hyarion committed:
Add missing bullet point

Co-authored-by: João Matos <joao@tritao.eu>

chennes commented:
It is probably worth noting that maintainers have the discretion to migrate to the next Ubuntu LTS up to several months ahead of its formal end-of-support-life. This is typically necessary because the GitHub CI system begins intentionally failing a random subset of runs using a soon-to-be EOL Ubuntu version, and we usually decide that the date that begins is the date we actually need to migrate.

hyarion commented:
@chennes ping?

chennes commented:
As far as I am concerned this effectively documents existing practice, with the addition of notification requirements for new dependencies (which is a good idea). So the next step in the FEP process is:

> When the author deems that the proposal is finalized and ready for discussion, it should be moved to Proposed state and the version should be changed to 1.0. A discussion should be created in the "FEP Discussion" section of the Discussion forum where the topic should have the same name as FEP and PR. The author may chose to create a new topic or ask Maintainers to simply move the topic from the "FEP Ideas" section. Once this has been accomplished, the PR containing the FEP should be merged into the repository assuming there are no formal problems with the PR.

hyarion committed:
Rename FEP-XXXX to FEP-0005 and simplify dependency policy

Simplified the dependency change process, added packaging-team
notification, and added CI runner brownout as an exceptional case.

hyarion commented:
@chennes I've updated the document to reflect our current process better.

chennes commented:
We should probably say something about the "contingency" state we're in right now -- although the current Ubuntu LTS includes Qt6, it *does not* include PySide6, which we have been essentially forced to pull onto that platform from another source so that we could begin dropping support for Qt5 (itself long-since EOL).

kadet1090 commented:
> We should probably say something about the "contingency" state we're in right now -- although the current Ubuntu LTS includes Qt6, it *does not* include PySide6, which we have been essentially forced to pull onto that platform from another source so that we could begin dropping support for Qt5 (itself long-since EOL).

For that we basically agreed that it must be possible to build and run, but can require additional PPA

hyarion committed:
Add contingency clause for dependencies missing from baseline LTS repos

Addresses feedback about the current PySide6 situation on Ubuntu 24.04,
where a PPA is needed. Allows additional package sources as a temporary
measure, provided FreeCAD remains buildable and runnable.

Lgt2x commented on FEPs/FEP-0005-process/README.md:
```diff
@@ -0,0 +1,98 @@
+
+### Dependency Changes
+
+* Any **new required dependency** or **change to the minimum required version** of an existing dependency should be communicated via a GitHub Pull Request labeled `Packaging/building`. The team `@FreeCAD/packaging-team` should be tagged to ensure visibility for downstream packagers.
```
This would be more effort, but creating a "packaging changelog" could be a good idea to help packagers catch up with a new release. Its content could be a simple bullet-point list of items like "changed minimum OCCT required version to X", "added dependency Y as a submodule, set variable `FREECAD_USE_EXTERNAL_Y` to override", etc. Packagers are not necessarily following the activity on GitHub.

hyarion replied:
We just created a packaging team on github to be able to ping packagers. Anyone is welcome to be pinged.

I think it would be great to have a machine readable list of our dependencies (min-max versions) that could be used in pixi and cmake. That way we'd have a single source of truth.

However, this FEP tries to document what is currently the policy, so I don't think we should add more to it if we're not already doing it.

Policies can change too, so adding it in later is an option.

Lgt2x commented on FEPs/FEP-0005-process/README.md:
```diff
@@ -0,0 +1,98 @@
+
+This policy defines the **baseline development environment** that FreeCAD must remain compatible with. It is distinct from the environment used to build official releases, which may use newer dependencies.
+
+* FreeCAD **must be buildable and runnable by contributors** on the **oldest Ubuntu LTS release still under standard support**.
```
A side effect not covered by this FEP that we figured out in PR https://github.com/FreeCAD/FreeCAD/pull/29535 is that effectively, new dependencies that have no package in ubuntu LTS somehow have to be vendored. It would be valuable to have a procedure to standardize it. Would this be in the scope of this FEP?

hyarion replied:
How to vender packages depends on the situation and project to vendor. Sometimes subtree is best, sometimes submodule and sometimes just copy the files might be the best option.

Adding a line that vendoring is an option could be good though, to make it explicit that it is an option.

adrianinsaval commented on FEPs/FEP-0005-process/README.md:
```diff
@@ -0,0 +1,98 @@
+
+This policy defines the **baseline development environment** that FreeCAD must remain compatible with. It is distinct from the environment used to build official releases, which may use newer dependencies.
+
+* FreeCAD **must be buildable and runnable by contributors** on the **oldest Ubuntu LTS release still under standard support**.
```
this either has poor wording or is not reflecting what we currently do or what I think we want to do, the oldest ubuntu LTS still under standard support is 22.04 but our current baseline is 24.04, and 26.04 is out now. In previous developers meetings many have asked to drop the oldest LTS and support only the latest 2 LTS, some even wanting only the latest.

hyarion replied:
You are correct, and I agree that supporting the oldest available is dumb, but it is our current policy as far as I know.
My idea here was to:
1. document our current policy, then
2. suggest changes to it.

@chennes or do you know if we've changed it while I was afk?

chennes replied:
At our [17 January 2026 developer meeting](https://github.com/FreeCAD/FreeCAD-developer-meetings/blob/main/Minutes/minutes-2026-01-17.md) it was decided that from 1.2dev on we would only support Ubuntu 24.04 LTS and later, with the additional caveat that we'd pull in PySide 6.4 on 24.04 using a PPA, since it wasn't available natively. The consensus was that the amount of developer effort to maintain support for Qt 5.15 was simply too large to be tenable.

We didn't make any blanket policy changes, just the single oldest-supported-Ubuntu-version change.

furgo16 replied:
I think the challenge here is that there is no PPA available with PySide6, as far as I could gather. That's why the CI script uses the KDE Neon overlay archive on top of the canonical Ubuntu archive packages, as it's the cleanest way to provide an up-to-date Qt6 that is in sync with PySide6. However, that is not a solution I feel we should ask users to adopt.

furgo16 commented on FEPs/FEP-0005-process/README.md:
```diff
@@ -0,0 +1,98 @@
+
+## Rationale
+
+Ubuntu LTS releases are widely used, well-supported, and provide a consistent foundation across both developer and user systems. Aligning with the oldest currently supported LTS (within its standard support period):
```
Thanks for the proposal. Some notes:

To me the "oldest currently supported LTS" constraint sounds:

- **arbitrary** - we are targetting developers who want to compile using CMake, not regular users. Surely those developers either use the most recent release, or if not, the most recent LTS release? I cannot think of a reason why most developers, even if they only use LTSs, they will not use the latest one.
- **limiting** - at this current point in time (Ubuntu 22.04 LTS is the oldest, 24.04 and 26.04 the newest), there's no optimal Qt 6 environment that matches what we are building FreeCAD with. Only on 26.04 are both Qt6 and PySide6 available from the Ubuntu archive. This means that developers using the seemingly FreeCAD-supported 22.04 release cannot compile out of the box without manually installing additional and non trivial dependencies (no PPA available for Qt6/PySide6). See the commits on https://github.com/FreeCAD/FreeCAD/pull/26894/commits for the history of hoops that needed to be jumped through to get CI to build on a native Ubuntu 24.04 image.

furgo16 replied:
I only realized after the comment that it's essentially piling on to https://github.com/FreeCAD/FreeCAD-Enhancement-Proposals/pull/13#discussion_r3177008605, apologies. The point remains, though.
